Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp

```diff
@@ -441,12 +444,44 @@ void registerWebKitGStreamerVideoEncoder()
         GStreamerRegistryScanner::singleton().refresh();
 }
 
+static Lock s_activePipelinesMapLock;
+static HashMap<CString, GRefPtr<GstElement>>& activePipelinesMap()
+{
+    static NeverDestroyed<HashMap<CString, GRefPtr<GstElement>>> activePipelines;
+    return activePipelines.get();
+}
+
+void registerActivePipeline(GRefPtr<GstElement>& pipeline)
+{
+    GUniquePtr<gchar> name(gst_object_get_name(GST_OBJECT_CAST(pipeline.get())));
+    Locker locker { s_activePipelinesMapLock };
+    activePipelinesMap().add(name.get(), GRefPtr<GstElement>(pipeline));
+}
+
+void unregisterPipeline(GRefPtr<GstElement>& pipeline)
+{
+    GUniquePtr<gchar> name(gst_object_get_name(GST_OBJECT_CAST(pipeline.get())));
+    Locker locker { s_activePipelinesMapLock };
+    activePipelinesMap().remove(name.get());
+}
+
 void deinitializeGStreamer()
 {
 #if ENABLE(MEDIA_STREAM)
     teardownGStreamerCaptureDeviceManagers();
 #endif
     teardownGStreamerRegistryScanner();
+
+    // Make sure there is no active pipeline left. Those might trigger deadlocks during gst_deinit().
+    {
+        Locker locker { s_activePipelinesMapLock };
+        for (auto& pipeline : activePipelinesMap().values()) {
+            GST_DEBUG("Pipeline %" GST_PTR_FORMAT " was left running. Forcing clean-up.", pipeline.get());
+            gst_element_set_state(pipeline.get(), GST_STATE_NULL);
```

Comment:
I would check if there is any error at the state change and assert on it.

Reply:
Transition to GST_STATE_NULL never fails, so not sure an assert is worth.